### ISSUE TITLE
issue when installing your module

### ISSUE THREAD

**Issue by supriyawishesu** [maxlinc/puppet-decrypt#9]
puppet module install puppet_decrypt
Notice: Preparing to install into /etc/puppetlabs/puppet/modules ...
Notice: Downloading from https://forgeapi.puppetlabs.com ...
Error: Could not execute operation for 'puppet_decrypt'
  The server being queried was /v3/releases?module=puppet_decrypt
  The HTTP response we received was '400 Bad Request'
    Check the author and module names are correct.

**Comment by maxlinc** [comment]
Here's the link to the module on puppetforge:
https://forge.puppetlabs.com/devopsy/puppet_decrypt

According to that, the install command needs my username:
`puppet module install devopsy-puppet_decrypt`
not
`puppet module install puppet_decrypt`

I've updated the readme.

**Comment by supriyawishesu** [comment]
Hi,
I am getting this issue. when I used your module
Error 400 on SERVER: Could not autoload puppet/parser/functions/decrypt:
cannot load such file -- encryptor on node ****
Let me know

**Comment by supriyawishesu** [comment]
Hi Also specifyied decrpt in my manifests

 $reposPassword   = "decrypt(hiera('wls_repo_password')"

##  more dot-fap-spr-t01.local.yaml

wls_repo_password: "ENC***********"

Let me know

**Comment by maxlinc** [comment]
It's because your server doesn't have the encryptor gem on the ruby load path. If your install as a module, the readme says:

> If installing as a module, you'll need to deal with the Gem prerequisites manually.
> `$ gem install encryptor`

You need to install that gem on all of your puppet agents, to the same ruby that's used for the puppet runs.

That's because there didn't seem to be a way to declare that a module depended on a gem when I developed this against puppet 2.7.14. There might be a better way to do that now (see comments [here](https://tickets.puppetlabs.com/browse/PUP-1879)).

I haven't been doing too much puppet work recently, though, so I probably won't be improving the install process unless someone sends a PR.

**Comment by maxlinc** [comment]
> `$reposPassword   = "decrypt(hiera('wls_repo_password')"`

I don't think it should be in quotes, because decrypt is a function.  It should probably be:

``` ruby
$reposPassword   = decrypt(hiera('wls_repo_password')
```

Check out [the hiera integration tests](https://github.com/maxlinc/puppet-decrypt/blob/master/features/hiera.feature) for working examples.

**Comment by supriyawishesu** [comment]
Hi,

I installed $ gem install encryptor on agent

But the still running into same issue.
Error: Could not retrieve catalog from remote server: Error 400 on SERVER:
Could not autoload puppet/parser/functions/decrypt: cannot load such file
-- encryptor on node dot-fap-spr-t01.local
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run

Help me

**Comment by maxlinc** [comment]
Not sure if I can help more, but I just release v0.2.0. Can you try and confirm you're still having a problem?

It might also help to have puppet install the gem (using [package](https://docs.puppetlabs.com/references/latest/type.html#package) with the gem provider) so you're sure it installs the same account and ruby install that puppet is using.

**Comment by iamteedoh** [comment]
I am working on setting up puppet-decrypt. I downloaded the following two files from rubygems.org:

> encryptor 
> puppet-decrypt

I have installed them both as follows since our servers are not allowed to communicate outside of our network:

> gem install /tmp/encryptor-1.3.0.gem 
> gem install /tmp/puppet-decrypt-0.2.0.gem

After installing, the documentation states that I need to add the $RUBYLIB directory to my profile. So, I have gone through root's .bash_profile and added the following entry:

> export RUBYLIB=/opt/puppet/lib:$RUBYLIB

After sourcing my profile, I try to verify that the gem is installed by running the following:

> puppet help crypt

However, I get the following error:

> [user1@testbed ~]# puppet help crypt 
> Error: Could not load help for the face crypt. 
> Please check the error logs for more information.
> Detail: "Could not find Puppet Face crypt"
> Error: Try 'puppet help help help' for usage

That error looks like it can't find the installation of encryptor or puppet-decrypt. However, if I check for the gems, I do see them in the system:

> [user1@testbed gems]# gem list --local | egrep "encryptor|decrypt" 
> encryptor (1.3.0) 
> puppet-decrypt (0.2.0)

What could I be missing to get this gem going in order to start encrypting data?

Resource links:

puppet-decrypt = https://rubygems.org/gems/puppet-decrypt 
encryptor = https://rubygems.org/gems/encryptor

**Comment by maxlinc** [comment]
@unix4linux - you said you made a change to root's .bashrc_profile, but the shell snippets show you're using user1, not root. That could be related to the problem.

I originally wrote this because at the time hiera-gpg was the only real option and I had workflows that I didn't think were a good match for hiera-gpg. I [wasn't the only one who wanted an alternative to hiera-gpg](http://slashdevslashrandom.wordpress.com/2013/06/03/my-griefs-with-hiera-gpg/), so there are now a few alternatives.

I recommend checking out [hiera-eyaml](https://github.com/TomPoulton/hiera-eyaml) first, because it looks like what I was trying to achieve with this project. I need to test it out a bit, but there's a good chance I may deprecate this project and direct people towards hiera-eyaml.

Another fairly new (but rather different) option is [blackbox](https://github.com/StackExchange/blackbox).

**Comment by iamteedoh** [comment]
@maxlinc Thank you for your response. the reason for user1 is because I was trying to sanitize my data ;). Sorry for the confusion there.

In regards to hiera-eyaml, I'll have to take a look at that some more. Looks like they are working on getting it into Puppet Forge :)

Thanks for sharing the alternatives and thank you for your contribution and your project.
